### ISSUE TITLE
Tensor __getitem__ not documented, sparse grad?

### ISSUE THREAD

**Issue by albertz** [pytorch/pytorch#101068]
### 📚 The doc issue

I was search for some documentation on `Tensor.__getitem__` but I don't find any, at least not [here](https://pytorch.org/docs/stable/tensors.html#torch.Tensor). Maybe it is elsewhere?

Esp, I wonder if it uses `torch.gather` internally when you give it some scalar index.

And I wonder if it uses `sparse_grad` for `torch.gather` or not.

I also wonder, why does `torch.gather` has such a `sparse_grad` option at all? Why is it not always sparse? Why is the default to have it dense? What are the downsides in always enabling this? The [`torch.gather` documentation](https://pytorch.org/docs/stable/generated/torch.gather.html) doesn't really say anything about this.

If `Tensor.__getitem__` does not use `torch.gather`, is the gradient sparse or not?

E.g. when I have a tensor `x` of shape [T, B, F], and I loop over T, would it be a good idea to have this code (when training)?
```python
for t in range(T):
   y = f(x[t])
   ...
```
If `__getitem__` is not using sparse gradients, wouldn't it be better to have sparse gradients in this case? Wouldn't it always be better to have sparse gradients?

### Suggest a potential alternative/fix

Add some documentation on `Tensor.__getitem__`.

Extend the documentation of `torch.gather` `sparse_grad` option.

Briefly address my questions here.

cc @alexsamardzic @nikitaved @pearu @cpuhrsch @amjames @bhosmer

**Comment by cpuhrsch** [comment]
@albertz - Currently index_select gradients are not always sparse. We've been wanting to work on a mode to control the layout of gradients, but haven't converged on an API yet. cc @albanD

**Comment by albertz** [comment]
And `x[index]` for some tensor `x` and some `index` has sparse gradients to `x`?

**Comment by cpuhrsch** [comment]
@albertz - No, `x[index]` will not produce sparse gradients either. It's not obvious or computationally cheap to know when it's better to use a sparse layout to store data vs. when not.

**Comment by cpuhrsch** [comment]
Currently you'd need to use our [custom autograd mechanism](https://pytorch.org/docs/stable/notes/extending.html#extending-torch-autograd) and implement your own variant of select to support sparse gradients.

**Comment by albanD** [comment]
> if you just hardcode to one case (sparse or dense) no matter the input, sparse grads would make more sense than dense grads, as in the majority of use cases, this is probably better. At least for x[index] and probably also torch.index_select. Or is this not right?

I wouldn't be as categorical for two reasons. First because our sparse support is not a broad as our Dense Tensor support. So this would lead to many models to fail on ops that don't have sparse support. And second, as Christian mentioned, it is not clear which sparse format would be best and if any of them will be. Especially because most of our users run this code on GPU and it needs very very high sparsity to be faster than dense when using usual sparsity formats.

**Comment by albanD** [comment]
Also the definition of indexing is the same as numpy which has a great documentation here: https://numpy.org/doc/stable/user/basics.indexing.html

**Comment by albertz** [comment]
Oh, by sparse gradient, you mean actually a sparse tensor in [COO](https://pytorch.org/docs/stable/sparse.html#sparse-coo-docs), [CSR](https://pytorch.org/docs/stable/sparse.html#sparse-csr-docs), [CSC](https://pytorch.org/docs/stable/sparse.html#sparse-csc-docs), [BSR](https://pytorch.org/docs/stable/sparse.html#sparse-bsr-docs), or [BSC](https://pytorch.org/docs/stable/sparse.html#sparse-bsc-docs) format?

I was expecting sth like [TensorFlow `IndexedSlices`](https://www.tensorflow.org/api_docs/python/tf/IndexedSlices), which is very simple in contrast, is a very direct representation of the inverse of `x[i]`, and thus is should be clearly the best representation at least for the grad of `x[i]`.

> The `IndexedSlices` class is used principally in the definition of gradients for operations that have sparse gradients (e.g. [`tf.gather`](https://www.tensorflow.org/api_docs/python/tf/gather)).

So I assume you don't have sth equivalent as `IndexedSlices` in PyTorch?

On compatibility with other ops, in TF, all ops would call `convert_to_tensor` on all arguments, and that would automatically convert it to a dense tensor. So no code would break by this. And code which can make efficient use of it need to check explicitly for the `IndexedSlices` type.

> Also the definition of indexing is the same as numpy which has a great documentation here: https://numpy.org/doc/stable/user/basics.indexing.html

I assumed this, but is it said anywhere that indexing is the same as NumPy? At least I did not find this when looking at the [`torch.Tensor` documentation](https://pytorch.org/docs/stable/tensors.html#torch.Tensor).

**Comment by albanD** [comment]
> So I assume you don't have sth equivalent as IndexedSlices in PyTorch?

You can do that pretty easily with a subclass of Tensor if you need (few examples here:https://github.com/albanD/subclass_zoo) . You can then define your custom backward that will return this object for you.
I would also note that there can be some questions on the semantic of such Tensor where the missing values are just 0s or "missing". You can check the `torch.masked` doc for more details on this semantic discussion.

**Comment by albertz** [comment]
> > So I assume you don't have sth equivalent as IndexedSlices in PyTorch?
> 
> You can do that pretty easily with a subclass of Tensor if you need (few examples here:https://github.com/albanD/subclass_zoo) . You can then define your custom backward that will return this object for you.

So wouldn't this be a good generic solution for PyTorch to use for the gradient of `x[i]`?

Then you can step-by-step add some more support for it, and it would not break anything.

In case of the gradients, the missing values would be just 0s.

**Comment by albanD** [comment]
Hey!

> So wouldn't this be a good generic solution for PyTorch to use for the gradient of x[i]?

I'm afraid it would not be backward compatible. Using such a subclass would be an observable change and would not be guaranteed to work if any downstream code expects a honest to goodness Tensor (for example some distributed collective or custom cpp code).
So I do think it is in the realm where it will work 90%+ of the time. But that still means it is BC-breaking and we need to be careful about such a change...
